### PR TITLE
Fix lsp weird behaviour on block cursors (#891)

### PR DESCRIPTION
**Pull request by acristoffers**
I believe I got it right this time. This not only fixes the server for block-cursors but also correctly detects builtins, `foo.@"bar"`-style field access and single-letter identifiers.

For line-cursor, the following changes will be perceived: anytime the cursor touches an identifier, it is considered to be on that identifier.

I changed the test file to include a lot more positions for the same contexts we had already, always following the idea that if it touches it is in.

The change boils down to

```zig
    var new_index = doc_index;
    if (new_index < text.len and isSymbolChar(text[new_index])) {
        new_index += 1;
    } else if (new_index + 1 < text.len and text[new_index] == '@') {
        new_index += 2;
    }
```

which, in plain english, (I hope it) means: 
- if the next char is a valid identifier char, include it.
- otherwise, if the next char is @ and we still have a char after that available in the buffer, add both.

### CODE REVIEW

**Review by Techatrix — DISMISSED** [review]
This change will cause auto completion to include a character to the right of the cursor which is nonsensical in my opinion. `getPositionContext` needs another parameter to control if an identifier to the right side should be included. This can then be disabled for completions and enabled for references, goto definition and hover.

Note that this change doesn't actually change the behavior of auto completions because there is not filtering based on the identifier at the end (yet).

**Comment by acristoffers** [comment]
But that would mean that you are requesting a completion in the middle of a word or right before a `@`, which is also nonsensical, no?

**Comment by Techatrix** [comment]
> But that would mean that you are requesting a completion in the middle of a word or right before a `@`, which is also nonsensical, no?

i don't really understand what you mean with this sentence (maybe i'm just bad at english).
Maybe this will clear things up:

auto completions:
`getPositionContext("std.<cursor>mem")` -> `"std."` 
`getPositionContext("std.me<cursor>m")` -> `"std.me"` 
`getPositionContext("std.<cursor>@"memcopy"")` -> `"std."`
`getPositionContext("std.@<cursor>"memcopy"")` -> `"std."`
`getPositionContext("std.@"<cursor>memcopy"")` -> `"std."`
`getPositionContext("std.@"me<cursor>mcopy"")` -> `"std.me"`
goto definition:
`getPositionContext("std.<cursor>mem")` -> `"std.mem"` 
`getPositionContext("std.me<cursor>m")` -> `"std.mem"` 
`getPositionContext("std.<cursor>@"memcopy"")` -> `"std.memcopy"` 
`getPositionContext("std.@<cursor>"memcopy"")` -> `"std.memcopy"`
`getPositionContext("std.@"<cursor>memcopy"")` -> `"std.memcopy"`
`getPositionContext("std.@"me<cursor>mcopy"")` -> `"std.memcopy"`

I think you will actually get something like `"std.@"memcopy"` or `"std.@"me"` but that can be fixed later.

**Comment by acristoffers** [comment]
I mean that basically all your examples of autocompletion make no sense. Why would you request completition in the middle of a word?

```zig
getPositionContext("std.mem.e<cursor>ql")
```

What is supposed to happen in this case? Rigt now I'm getting this:

image

If I chose the first options the result is

image

and if I chose the second option it becomes

image

So no matter what I chose, the editor simply inserts text in the middle, so now I have to go to the end and delete it. So I don't see why it even makes sense to request completion in the middle of a word.

I may be missing the use case for this here, but it makes no sense to me to consider this case as something that needs to be taken into consideration because it doesn't make sense to start with.

**Comment by Techatrix** [comment]
It can make sense to request completion in the middle of a word. 

The fact that zls will insert the entire function signature instead of removing the characters after the cursor is the part that makes no sense. This should be fixed but that a separate issue. The changes i request are just so that we don't need to redo this later.

**Review by SuperAuguste — APPROVED** [review]
